File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/22-PQ/LeetCode/LeetCode-0023.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
#include <queue>

class Solution {
public:
    // This is the efficient O(nlogk) algorithm
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.size() == 0) return NULL;
        if (lists.size() == 1) return lists[0];

        auto cmp = [](pair<int, int> &left, pair<int, int> &right) {
            return (left.first >= right.first);
        };
	    priority_queue<pair<int, int>, vector<pair<int, int>>,
                       decltype(cmp)> H(cmp);

        // Create the initial heap
        for (size_t i=0; i<lists.size(); i++){
            if (lists[i] != NULL){
                H.push(make_pair(lists[i]->val, i));
            } //end-if
        } //end-for

        ListNode *head = NULL, **tail = &head;

        while (!H.empty()){
            auto [_, index] = H.top();
            H.pop();

            *tail = lists[index];
            tail = &(*tail)->next;
            lists[index] = lists[index]->next;

            if (lists[index]){
                H.push({lists[index]->val, index});
            } //end-if
        } //end-while

        return head;
    } //end-mergeKLists
};

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/22-PQ/LeetCode/LeetCode-0023.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0) return null;
        if (lists.length == 1) return lists[0];

        record Pair(int val, int listIndex){};
        PriorityQueue<Pair> Q = 
            new PriorityQueue<>((left, right)->left.val()-right.val());

        // Create the initial PQ
        for (int i=0; i<lists.length; i++){
            if (lists[i] != null)
                Q.offer(new Pair(lists[i].val, i));
        } //end-for

        ListNode head = null, tail = null;

        while (!Q.isEmpty()){
            var index = Q.poll().listIndex();

            if (tail == null){ 
                head = tail = lists[index];
            } else {
                tail.next = lists[index];
                tail = tail.next;
            } // end-else

            lists[index] = lists[index].next;

            if (lists[index] != null)
                Q.offer(new Pair(lists[index].val, index));
        } //end-while

        return head;
    }
}
